exec/cnex/number.c

```diff
@@ -17,49 +17,96 @@
 {
     static char buf[50];
-    char val[50];
-    bid128_to_string(val, x);
-    sprintf(buf, "%g", atof(val));
+    number_toString(x, buf, sizeof(buf));
     return buf;
 }
 
-/* ToDo: Implement proper number to string formatter. */
 void number_toString(Number x, char *buf, size_t len)
 {
-    //const int PRECISION = 34; /* ToDo: Finish implementation */
-    int iLeadingZeros = 0;
+    memset(buf, 0x00, len);
+    const int PRECISION = 34;
 
     assert(len != 0);
 
     char val[50];
-    char lead[50];
     bid128_to_string(val, x);
 
     char *s = &val[0];
 
-    /* Skip over the leading +, if there is one. */
-    while(*s != '\0' && (*s == '+')) s++;
+    /* Skip over the leading sign value, if there is one. */
+    while(*s != '\0' && (*s == '+') || (*s == '-')) s++;
 
     char *v = s;
 
     while (*s != '\0' && (*s != 'E')) s++;
 
     char *E = s;
 
-    if (*s == 'E') {
-        /* Deal with the exponent. */
-        int i;
-        iLeadingZeros = abs(atoi(++s));
-        for (i = 0; i < iLeadingZeros; i++) {
-            lead[i] = '0';
-        }
-        lead[i++] = '.';
-        lead[i++] = '\0';
-        strcpy(buf, lead);
+    if (*s != 'E') {
+        buf = val;
+        return;
     }
 
+    int exponent = atoi(++s);
     *E = '\0';
-    strcat(buf, v);
+    char *p = E - 1;
+
+    while ((p > v) && (*p == '0')) p--;
+
+    const int last_significant_digit = (int)(p - v);
+    //if (last_significant_digit == 0) {
+    //    buf[0] = '0';
+    //    buf[1] = '\0';
+    //    return;
+    //}
+    const int trailing_zeros = (int)strlen(v) - (last_significant_digit + 1);
+    exponent += trailing_zeros;
+    char sbuf[50];
+    char r[50] = { 0 };
+    memcpy(sbuf, v, last_significant_digit + 1);
+    sbuf[last_significant_digit+1] = '\0';
+    int slen = (int)strlen(sbuf);
+
+    if (exponent != 0) {
+        if (exponent > 0 && slen + exponent <= PRECISION+1) {
+            memcpy(r, sbuf, slen);
+            for (int i = slen; i < (slen + exponent); i++) {
+                r[i] = '0';
+            }
+        } else if (exponent < 0 && -exponent < slen) {
+            memcpy(r, &sbuf[0], slen+exponent);
+            r[slen+exponent] = '.';
+            memcpy(&r[slen+exponent]+1, &sbuf[slen+exponent], slen-exponent);
+        } else if (exponent < 0 && -exponent == slen-1) {
+            memcpy(r, "0.", 2);
+            memcpy(&r[2], sbuf, slen);
+        } else if (exponent < 0 && slen - exponent <= PRECISION+2) {
+            int count = 2;
+            memcpy(r, "0.", count);
+            for (int i = 0; i < -exponent-slen;i++, count++) {
+                r[count] = '0';
+            }
+            memcpy(&r[count], sbuf, slen);
+            r[count+slen] = '\0';
+        } else {
+            exponent += slen - 1;
+            if (slen >= 3) {
+                memcpy(r, sbuf, 1);
+                r[1] = '.';
+                memcpy(&r[2], &sbuf[1], slen-1);
+            }
+            char exp[36];
```

Comment:
This was just to make very certain I'd have enough room for the exponent, and is basically borrowed from the constant in the source for the precision.  What is a better, safer value to use here?  4?  Since the exponent will (should) never be larger than e99, correct?

Reply:
No, the exponent can be quite a bit larger than 99 (this is bid128, not a handheld calculator). See https://neon-lang.org/docs/reference/types.html#number for the full floating point range.